plugins/libplugin-pay.c

```diff
@@ -0,0 +1,780 @@
+			json_add_hex_talarr(req->js, "payload", payload);
+		}else {
+			tlv = tal_arr(tmpctx, u8, 0);
+			towire_tlvstream_raw(&tlv, hop->tlv_payload->fields);
+			payload = tal_arr(tmpctx, u8, 0);
+			towire_bigsize(&payload, tal_bytelen(tlv));
+			towire(&payload, tlv, tal_bytelen(tlv));
+			json_add_hex_talarr(req->js, "payload", payload);
+			tal_free(tlv);
+		}
+		tal_free(payload);
+		json_object_end(req->js);
+	}
+	json_array_end(req->js);
+
+	json_add_hex_talarr(req->js, "assocdata",
+			    p->createonion_request->assocdata);
+
+	if (p->createonion_request->session_key)
+		json_add_secret(req->js, "sessionkey",
+				p->createonion_request->session_key);
+
+	send_outreq(p->plugin, req);
+}
+
+/* Mutual recursion. */
+static struct command_result *payment_finished(struct payment *p);
+
+/* A payment is finished if a) it is in a final state, of b) it's in a
+ * child-spawning state and all of its children are in a final state. */
+static bool payment_is_finished(struct payment *p)
+{
+	bool running_children = false;
+	if (p->step == PAYMENT_STEP_FAILED || p->step == PAYMENT_STEP_SUCCESS)
+		return true;
+	else if(p->step == PAYMENT_STEP_SPLIT || p->step == PAYMENT_STEP_RETRY) {
+		for (size_t i = 0; i < tal_count(p->children); i++)
+			running_children |= !payment_is_finished(p->children[i]);
+		return !running_children;
+	} else {
+		return false;
+	}
+}
+
+static enum payment_step payment_aggregate_states(struct payment *p)
+{
+	enum payment_step agg = p->step;
+
+	for (size_t i=0; i<tal_count(p->children); i++)
+		agg |= payment_aggregate_states(p->children[i]);
+
+	return agg;
+}
+
+/* A payment is finished if a) it is in a final state, of b) it's in a
+ * child-spawning state and all of its children are in a final state. */
+static bool payment_is_success(struct payment *p)
+{
+	return (payment_aggregate_states(p) & PAYMENT_STEP_SUCCESS) != 0;
+}
+
+/* Function to bubble up completions to the root, which actually holds on to
+ * the command that initiated the flow. */
+static struct command_result *payment_child_finished(struct payment *p,
+						     struct payment *child)
+{
+	if (!payment_is_finished(p))
+		return command_still_pending(NULL);
+
+	/* Should we continue bubbling up? */
+	return payment_finished(p);
+}
+
+/* This function is called whenever a payment ends up in a final state, or all
+ * leafs in the subtree rooted in the payment are akk in a final state. It is
+ * called only once, and it is guaranteed to be called in post-order
+ * traversal, i.e., all children are finished before the parent is called. */
+static struct command_result *payment_finished(struct payment *p)
+{
+	struct payment_tree_result result = payment_collect_result(p);
+	struct json_stream *ret;
+	struct command *cmd = p->cmd;
+
+	p->end_time = time_now();
+
+	/* Either none of the leaf attempts succeeded yet, or we have a
+	 * preimage. */
+	assert((result.treestates & PAYMENT_STEP_SUCCESS) == 0 ||
```

Comment:
your comment says 'leaf attempts' but you're checking the 'treestates' here?

Reply:
Ouch, good catch :+1: